### PR TITLE
EDURED106

### PR DESCRIPTION
**Pull request by risinggalaxy**

### CODE REVIEW

File: Sources/EduIDExpansion/Flows/CreateEduID/ViewModels/CreatePincodeAndBiometricAccessViewModel.swift [hunk 1]
```diff
@@ -84,37 +78,47 @@ final class CreatePincodeAndBiometricAccessViewModel: NSObject {
...
             do{
                 let enrolment = try await TiqrControllerAPI.startEnrollmentWithRequestBuilder()
-                    .addHeader(name: Constants.Headers.authorization, value: keychain.getString(for: Constants.KeyChain.accessToken))
+                    .addHeader(name: Constants.Headers.authorization, value: keychain.getString(for: Constants.KeyChain.accessToken) ?? "")
                     .execute()
                     .body
                 
                 ServiceContainer.sharedInstance().challengeService.startChallenge(fromScanResult: enrolment.url ?? "") { [weak self] type, object, error in
                     guard let self else { return }
+                    self.secondEnteredPin.removeLast(2)
```
Comment (dzolnai) [inline Sources/EduIDExpansion/Flows/CreateEduID/ViewModels/CreatePincodeAndBiometricAccessViewModel.swift:87]: Why is this required?

File: Sources/EduIDExpansion/Flows/CreateEduID/ViewModels/CreatePincodeAndBiometricAccessViewModel.swift [hunk 2]
```diff
@@ -84,37 +78,47 @@ final class CreatePincodeAndBiometricAccessViewModel: NSObject {
...
         biometricService.useOnDeviceBiometricFeature { [weak self] success, error in
             guard let self else { return }
             if success {
-                (viewController.biometricApprovaldelegate as? CreateEduIDViewControllerDelegate)?.createEduIDViewControllerShowNextScreen(viewController: viewController)
+                if let enrolment = self.enrollmentChallenge {
+                    if let managedObject = enrolment.identity.managedObjectContext {
+                        enrolment.identity.biometricIDEnabled = NSNumber(value: 1)
+                        do {
+                            try managedObject.save()
+                            (viewController.biometricApprovaldelegate as? CreateEduIDViewControllerDelegate)?.createEduIDViewControllerShowNextScreen(viewController: viewController)
```
Comment (dzolnai) [inline Sources/EduIDExpansion/Flows/CreateEduID/ViewModels/CreatePincodeAndBiometricAccessViewModel.swift]: Please use protocols

File: Sources/EduIDExpansion/Flows/Activity/ActivityViewController.swift [hunk 3]
```diff
@@ -26,16 +26,16 @@ class ActivityViewController: BaseViewController {
...
-        viewModel.loadData()
-        
     }
     
     override func viewWillAppear(_ animated: Bool) {
         super.viewWillAppear(animated)
-        
         screenType.configureNavigationItem(item: navigationItem, target: self, action: #selector(dismissActivityScreen))
+        DispatchQueue.main.async { [weak self] in
+            guard let self else { return }
```
Comment (dzolnai) [inline Sources/EduIDExpansion/Flows/Activity/ActivityViewController.swift]: Why does this need to be async?

**Review by dzolnai — CHANGES_REQUESTED** [review]

File: Sources/EduIDExpansion/Flows/CreateEduID/ViewModels/CreatePincodeAndBiometricAccessViewModel.swift (hunk 2 above)
Comment (risinggalaxy) [reply Sources/EduIDExpansion/Flows/CreateEduID/ViewModels/CreatePincodeAndBiometricAccessViewModel.swift:116]: there may be scenarios where an existing user has a challange stored on their device, but they don't have an AccessToken, so if that's the case the app will open the browser to get an Access Token, I noticed the UI doesn't get updated when users are redirected from a browser to the app.

File: Sources/EduIDExpansion/Flows/Activity/ActivityViewController.swift (hunk 3 above)
Comment (risinggalaxy) [reply Sources/EduIDExpansion/Flows/Activity/ActivityViewController.swift]: there may be a scenario where an existing Tiqr user has an account but doesn't have an AccessToken, so the app opens the browser to retrieve an AccessToken, then the user is redirected to the app, I noticed the UI didn't update when user came back from the browser. also the loadData() method is an async task.

File: Sources/EduIDExpansion/Flows/CreateEduID/ViewModels/CreatePincodeAndBiometricAccessViewModel.swift (hunk 2 above)
Comment (risinggalaxy) [reply Sources/EduIDExpansion/Flows/CreateEduID/ViewModels/CreatePincodeAndBiometricAccessViewModel.swift]: can you explain it further?

File: Sources/EduIDExpansion/Flows/CreateEduID/ViewModels/CreatePincodeAndBiometricAccessViewModel.swift (hunk 1 above)
Comment (risinggalaxy) [reply Sources/EduIDExpansion/Flows/CreateEduID/ViewModels/CreatePincodeAndBiometricAccessViewModel.swift:87]: the previous developer implemented the same logic for both SMS verification which uses 6 digits and for the pin verification which requires 4 digits, I noticed rewriting the entire thing needed a bit more time and focus, since the project already is delayed for so long, I thought I'll rewrite it once there is a better chance for it. so right now I just drop the last 2 digits so I can save a 4 digit pin code.

**Review by dzolnai — APPROVED** [review]
